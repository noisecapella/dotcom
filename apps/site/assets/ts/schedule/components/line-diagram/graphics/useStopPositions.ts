import { useDispatch } from "react-redux";
import { useRef, useEffect, useCallback } from "react";
import { LineDiagramStop } from "../../__schedule";
import { isMergeStop } from "../line-diagram-helpers";
import { BASE_LINE_WIDTH, BRANCH_SPACING } from "./graphic-helpers";

export interface RefList {
  [key: string]: React.MutableRefObject<HTMLElement | null>;
}

function useStopRef(prev: RefList, stop: LineDiagramStop): RefList {
  return {
    ...prev,
    [stop.route_stop.id]: useRef<HTMLElement>(null)
  };
}

const stopById = (stopId: string): ((stop: LineDiagramStop) => boolean) =>
  // eslint-disable-next-line camelcase
  ({ route_stop }) => stopId === route_stop.id;

const xCoordForStop = (stop: LineDiagramStop): number => {
  if (isMergeStop(stop)) {
    return BASE_LINE_WIDTH + 1;
  }
  // usually just a function of stop_data length
  return BRANCH_SPACING * (stop.stop_data.length - 1) + BASE_LINE_WIDTH + 1;
};

export default function useStopPositions(
  stops: LineDiagramStop[]
): [RefList, () => void] {
  const stopRefsMap: RefList = stops.reduce(useStopRef, {});
  const dispatchStopCoords = useDispatch();
  const updateAllStops = useCallback((): void => {
    Object.entries(stopRefsMap).forEach(([stopId, ref]) => {
      const x = xCoordForStop(stops.find(stopById(stopId))!);
      let coordinates = null;
      if (ref && ref.current) {
        const { offsetTop, offsetHeight } = ref.current;
        const y = offsetTop + offsetHeight / 2;
        coordinates = [x, y];
      }
      dispatchStopCoords({
        type: "set",
        stop: stopId,
        coords: coordinates
      });
    });
  }, [dispatchStopCoords, stopRefsMap, stops]);

  useEffect(() => {
    updateAllStops();
    window.addEventListener("resize", updateAllStops);
    return () => window.removeEventListener("resize", updateAllStops);
  }, [updateAllStops]);

  return [stopRefsMap, updateAllStops];
}
