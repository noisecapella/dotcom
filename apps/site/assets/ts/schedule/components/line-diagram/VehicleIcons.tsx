import React from "react";
import { renderToString } from "react-dom/server";
import { useSelector } from "react-redux";
import { RouteStop, LineDiagramVehicle } from "../__schedule";
import CrowdingPill from "./CrowdingPill";
import { TooltipWrapper, vehicleArrowIcon } from "../../../helpers/icon";
import { StopCoord, CoordState, CIRC_RADIUS } from "./graphics/graphic-helpers";

interface VehicleIconsProps {
  stop: RouteStop;
  vehicles: LineDiagramVehicle[] | null;
}

const CrowdingIconString = (vehicle: LineDiagramVehicle): string =>
  renderToString(<CrowdingPill crowding={vehicle.crowding} />);

const VehicleIcons = ({
  stop,
  vehicles
}: VehicleIconsProps): JSX.Element | null => {
  const coords: StopCoord | null = useSelector(
    (state: CoordState) => state[stop.id]
  );
  if (!vehicles || !coords) return null;
  const [x, y] = coords;
  const left = `${x - CIRC_RADIUS}px`;
  const tooltips = vehicles.map(vehicle => {
    const top = `${{
      // eslint-disable-next-line camelcase
      in_transit: y - 50,
      incoming: y - 25,
      stopped: y - 10
    }[vehicle.status] || y}px`;

    return (
      <div
        key={vehicle.id}
        className="m-schedule-diagram__vehicle"
        style={{ top, left }}
      >
        <TooltipWrapper
          tooltipText={`<div class="m-schedule-diagram__vehicle-tooltip">${
            vehicle.crowding ? `${CrowdingIconString(vehicle)}<br/>` : ""
          }${vehicle.tooltip}</div>`}
          tooltipOptions={{ placement: "right", animation: false, html: true }}
        >
          {vehicleArrowIcon("m-schedule-diagram__vehicle--icon")}
        </TooltipWrapper>
      </div>
    );
  });

  return <>{tooltips}</>;
};

export default VehicleIcons;
