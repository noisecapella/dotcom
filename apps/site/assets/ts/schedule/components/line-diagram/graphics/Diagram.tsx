import { max } from "lodash";
import React, { ReactElement } from "react";
import { LineDiagramStop, LineDiagramVehicle } from "../../__schedule";
import {
  hasBranchLines,
  isBranchTerminusStop,
  areOnDifferentBranchLines,
  diagramWidth,
  isMergeStop
} from "../line-diagram-helpers";
import Line from "./Line";
import Merges from "./Merge";
import Stop from "./Stop";
import { routeToModeName } from "../../../../helpers/css";
import { LiveDataByStop } from "../__line-diagram";
import VehicleIcons from "../VehicleIcons";
import { getCurrentState } from "../../../store/ScheduleStore";
import { DirectionId, HeadsignWithCrowding } from "../../../../__v3api";
import { isAGreenLineRoute } from "../../../../models/route";
import { BASE_LINE_WIDTH, DiagonalHatchPattern } from "./graphic-helpers";

interface DiagramProps {
  lineDiagram: LineDiagramStop[];
  liveData: LiveDataByStop | null;
}

const branchingDescription = (lineDiagram: LineDiagramStop[]): string => {
  const mergeStops = lineDiagram.filter(isMergeStop);
  const branchText = mergeStops.map(mergeStop => {
    const mergeStopName = mergeStop.route_stop.name;
    const branchNames = mergeStop.stop_data
      .filter(sd => sd.branch)
      .map(sd => sd.branch);
    return `Use ${mergeStopName} to transfer to trains on ${branchNames.join(
      " or "
    )} branches. `;
  });
  return branchText.join("");
};
const diagramDescription = (
  lineDiagram: LineDiagramStop[],
  direction: DirectionId
): string => {
  const text = `${lineDiagram.length} stops`;
  const {
    direction_destinations: destinations,
    direction_names: names
  } = lineDiagram[0].route_stop.route!;
  if (destinations) {
    return `${text} going ${names[direction]} to ${destinations[direction]}`;
  }
  return text;
};

interface VehicleIconSetProps {
  stop: LineDiagramStop;
  liveData: LiveDataByStop | null;
}

const LiveVehicleIconSet = ({
  stop,
  liveData
}: VehicleIconSetProps): ReactElement<HTMLElement> | null => {
  const stopId = stop.route_stop.id;
  if (!liveData || !liveData[stopId]) return null;
  // Hide vehicles arriving to the origin from 'off the line'
  const vehicleData: LineDiagramVehicle[] = [];

  const headsignsForStop = liveData[stopId];

  return (
    <VehicleIcons
      key={`${stopId}-vehicles`}
      stop={stop.route_stop}
      vehicles={vehicleData}
      headsigns={headsignsForStop as HeadsignWithCrowding[]}
    />
  );
};

const Diagram = (props: DiagramProps): ReactElement<HTMLElement> | null => {
  const { lineDiagram, liveData } = props;
  const { selectedDirection } = getCurrentState();
  const width = diagramWidth(
    max(lineDiagram.map(ld => ld.stop_data.length)) || 1
  );

  return (
    <>
      {lineDiagram.map(stop => (
        <LiveVehicleIconSet
          key={stop.route_stop.id}
          stop={stop}
          liveData={liveData}
        />
      ))}
      <svg
        xmlns="http://www.w3.org/2000/svg"
        role="img"
        aria-labelledby="diagram-title diagram-desc"
        className={`line-diagram-svg ${routeToModeName(
          lineDiagram[0].route_stop.route!
        )}`}
        width={`${width + 4}px`}
        height="100%"
        style={{ left: BASE_LINE_WIDTH / 2 }}
      >
        <title id="diagram-title">
          Line diagram for{" "}
          {isAGreenLineRoute(lineDiagram[0].route_stop.route!)
            ? "Green Line"
            : lineDiagram[0].route_stop.route!.name}
        </title>
        <desc id="diagram-desc">
          {diagramDescription(lineDiagram, selectedDirection)}
          {hasBranchLines(lineDiagram) && branchingDescription(lineDiagram)}
        </desc>

        <defs>{DiagonalHatchPattern}</defs>
        {/* If there are multiple branches, draw the lines and curves for those */
        hasBranchLines(lineDiagram) && <Merges lineDiagram={lineDiagram} />}

        {/* Draw lines between stops */
        lineDiagram.map((current, idx, self) => {
          if (self[idx + 1]) {
            if (
              isBranchTerminusStop(current) &&
              areOnDifferentBranchLines(current, self[idx + 1])
            )
              return null; // don't draw a line from a branch end to the main line

            return (
              <Line
                key={`${current.route_stop.id}-${self[idx + 1].route_stop.id}`}
                from={current}
                to={self[idx + 1]}
              />
            );
          }
          return null; // this is the end of the line, nothing to draw a line to
        })}

        {/* Draw circles for each stop */
        lineDiagram.map(stop => (
          <Stop key={stop.route_stop.id} stop={stop} />
        ))}
      </svg>
    </>
  );
};

export default Diagram;
