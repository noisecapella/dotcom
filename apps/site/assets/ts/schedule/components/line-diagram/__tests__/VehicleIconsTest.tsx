import React from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import {
  LineDiagramVehicle,
  RouteStop,
  LineDiagramStop,
  RouteStopRoute
} from "../../__schedule";
import VehicleIcons from "../VehicleIcons";
import { createLineDiagramCoordStore } from "../graphics/graphic-helpers";
import {
  Prediction,
  PredictedOrScheduledTime,
  PredictedOrScheduledTimeWithCrowding,
  HeadsignWithCrowding
} from "../../../../__v3api";

// mock the redux state
jest.spyOn(redux, "useSelector").mockImplementation(selector =>
  selector({
    "test-stop": [12, 80]
  })
);
const stop = { id: "test-stop", name: "Test Stop" } as RouteStop;
const vehicles = [
  { id: "vehicle-1", status: "in_transit", tooltip: "vehicle 1 tooltip text" },
  { id: "vehicle-2", status: "incoming", tooltip: "vehicle 2 tooltip text" },
  { id: "vehicle-3", status: "stopped", tooltip: "vehicle 3 tooltip text" }
] as LineDiagramVehicle[];
const store = createLineDiagramCoordStore([
  { route_stop: stop } as LineDiagramStop
]);

const tooltipText = (wrapper: ReactWrapper) =>
  wrapper.find("[tooltipText]").prop("tooltipText");

describe("VehicleIcons with no vehicles", () => {
  it("doesn't render", () => {
    const wrapper = mount(
      <redux.Provider store={store}>
        <VehicleIcons stop={stop} vehicles={null} />
      </redux.Provider>
    );

    expect(wrapper.html()).toBeFalsy();
  });
});

describe("VehicleIcons with vehicles", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <VehicleIcons stop={stop} vehicles={vehicles} />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders each vehicle", () => {
    const vehicleNodes = wrapper.find(".m-schedule-diagram__vehicle");
    expect(vehicleNodes).toHaveLength(vehicles.length);
  });

  it.each`
    index | tooltip
    ${0}  | ${"vehicle 1 tooltip text"}
    ${1}  | ${"vehicle 2 tooltip text"}
    ${2}  | ${"vehicle 3 tooltip text"}
  `("positions vehicles according to status $status", ({ index, tooltip }) => {
    const node = wrapper.find(".m-schedule-diagram__vehicle").at(index);
    expect(tooltipText(node)).toContain(tooltip);
  });

  it.each`
    index | expectedPosition | status
    ${0}  | ${30}            | ${"in_transit"}
    ${1}  | ${55}            | ${"incoming"}
    ${2}  | ${70}            | ${"stopped"}
  `(
    "positions vehicles according to status $status",
    ({ index, expectedPosition }) => {
      const node = wrapper.find(".m-schedule-diagram__vehicle").at(index);
      const { top } = node.get(0).props.style; // e.g. "30px"
      const top_number = parseInt(top.substring(0, 2)); // e.g. 30
      expect(top_number).toEqual(expectedPosition);
    }
  );
});

it("VehicleIcons includes the vehicle crowding status if available", () => {
  const wrapper = mount(
    <redux.Provider store={store}>
      <VehicleIcons
        stop={stop}
        vehicles={[
          {
            id: "v1",
            status: "incoming",
            crowding: "some_crowding",
            tooltip: "tooltip text"
          }
        ]}
      />
    </redux.Provider>
  );
  expect(tooltipText(wrapper)).toContain("Some crowding");
});
