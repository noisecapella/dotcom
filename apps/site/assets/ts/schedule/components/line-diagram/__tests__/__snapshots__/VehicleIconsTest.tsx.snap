// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleIcons with vehicles renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <VehicleIcons stop={{...}} vehicles={{...}}>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 1 tooltip text</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 1 tooltip text</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 1 tooltip text</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 2 tooltip text</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 2 tooltip text</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 2 tooltip text</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
    <div className=\\"m-schedule-diagram__vehicle\\" style={{...}}>
      <TooltipWrapper tooltipText=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 3 tooltip text</div>\\" tooltipOptions={{...}}>
        <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"right\\" data-animation={false} data-html={true} data-selector=\\"true\\" data-original-title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 3 tooltip text</div>\\" title=\\"<div class=\\"m-schedule-diagram__vehicle-tooltip\\">vehicle 3 tooltip text</div>\\">
          <span className=\\"notranslate m-schedule-diagram__vehicle--icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
        </span>
      </TooltipWrapper>
    </div>
  </VehicleIcons>
</Provider>"
`;
